catalog/service/rest/src/main/java/org/projectnessie/catalog/service/rest/IcebergS3SignParams.java

```diff
@@ -0,0 +1,193 @@
+/*
+ * Copyright (C) 2024 Dremio
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.projectnessie.catalog.service.rest;
+
+import static org.projectnessie.catalog.formats.iceberg.rest.IcebergError.icebergError;
+import static org.projectnessie.catalog.formats.iceberg.rest.IcebergS3SignResponse.icebergS3SignResponse;
+
+import io.smallrye.mutiny.Multi;
+import io.smallrye.mutiny.Uni;
+import jakarta.ws.rs.core.Response.Status;
+import java.net.URI;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.CompletionStage;
+import org.immutables.value.Value;
+import org.projectnessie.api.v2.params.ParsedReference;
+import org.projectnessie.catalog.files.api.RequestSigner;
+import org.projectnessie.catalog.files.api.SigningRequest;
+import org.projectnessie.catalog.files.api.SigningResponse;
+import org.projectnessie.catalog.files.s3.S3BucketOptions;
+import org.projectnessie.catalog.files.s3.S3Utils;
+import org.projectnessie.catalog.formats.iceberg.rest.IcebergException;
+import org.projectnessie.catalog.formats.iceberg.rest.IcebergS3SignRequest;
+import org.projectnessie.catalog.formats.iceberg.rest.IcebergS3SignResponse;
+import org.projectnessie.catalog.model.snapshot.NessieEntitySnapshot;
+import org.projectnessie.catalog.service.api.CatalogService;
+import org.projectnessie.catalog.service.api.SnapshotReqParams;
+import org.projectnessie.catalog.service.api.SnapshotResponse;
+import org.projectnessie.error.NessieContentNotFoundException;
+import org.projectnessie.error.NessieNotFoundException;
+import org.projectnessie.model.Content;
+import org.projectnessie.model.ContentKey;
+import org.projectnessie.model.IcebergTable;
+import org.projectnessie.model.IcebergView;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@Value.Immutable
+abstract class IcebergS3SignParams {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(IcebergS3SignParams.class);
+
+  abstract IcebergS3SignRequest request();
+
+  abstract ParsedReference ref();
+
+  abstract ContentKey key();
+
+  abstract String baseLocation();
```

Comment:
Does it make sense to add a check that this is a "pure" `s3://` location (and not s3a/s3n)?

Reply:
OK for now – long term I was rather leaning the idea of "pruning" at the warehouse config level. Basically all the locations are derived from that, so if you prune that, all the rest is pruned too.

Reply:
We can't do that there. The `a`+`n` could have a special meaning to clients in the way they read & write data.